cjwkernel/pandas/types.py

```diff
@@ -649,49 +702,87 @@ def arrow_table_to_dataframe(
 
 
-SimpleModuleError = Union[str, ModuleI18nMessage]
-ModuleErrorWithQuickFixes = Union[
-    SimpleModuleError, Dict[str, Union[SimpleModuleError, List[Any]]]
-]
-# The dict must have keys `"message": SimpleModuleError` and `"quickFixes": List[Any]`. We can define it in python 3.8 using `TypedDict`s
-ModuleError = Union[ModuleErrorWithQuickFixes, List[ModuleErrorWithQuickFixes]]
-""" ModuleError is the type we expect from modules to return when there is an error.
-When coercing, we convert it to the simpler type `ProcessResultError` held in `ProcessResult`.
-"""
-ProcessResultError = List[Tuple[atypes.I18nMessageDict, List[QuickFix]]]
-
-
-def _coerce_to_process_result_error(
-    value: ModuleError, can_nest: bool = True
-) -> ProcessResultError:
-    if not value and can_nest:
-        return []
-    elif isinstance(value, (str, tuple)):
-        return [(_coerce_to_i18n_message_dict(value), [])]
-    elif isinstance(value, dict):
-        if "message" not in value or "quickFixes" not in value:
-            raise ValueError("Missing 'message' or 'quickFixes' in %s" % value)
+@dataclass(frozen=True)
+class ProcessResultError:
+    message: I18nMessage
+    quick_fixes: List[QuickFix] = field(default_factory=list)
+
+    @classmethod
+    def coerce_list(
+        cls, value: Optional[mtypes.ModuleErrorsWithQuickFixes]
+    ) -> List[ProcessResultError]:
+        """Convert a list of errors as returned by module to a list of members of this dataclass.
+        
+        Raises:
+        - TypeError, if the value is not a list
+        - ValueError, if some element of the list cannot be coerced to a member of this dataclass
+        """
+        if value is None:
+            return []
+        elif not isinstance(value, list):
+            raise TypeError("Expected a list, got %s" % type(value).__name__)
         else:
-            return [
-                (
-                    _coerce_to_i18n_message_dict(value["message"]),
-                    [QuickFix.coerce(qf) for qf in value["quickFixes"]],
-                )
-            ]
-    elif (isinstance(value, list)) and can_nest:
-        try:
-            # collect the results for each item in a single list of tuples
-            return list(
-                chain.from_iterable(
-                    [_coerce_to_process_result_error(val, False) for val in value]
-                )
+            try:
+                # collect the results for each item in a single list of tuples
+                return [cls.coerce(val) for val in value]
+            except (ValueError, TypeError) as e:
+                raise ValueError(
+                    "The list %s cannot be coerced to module error" % value
+                ) from e
+
+    @classmethod
+    def coerce(cls, value: mtypes.ModuleErrorWithQuickFixes) -> ProcessResultError:
+        """Convert an error as returned by module to a member of this dataclass.
+        
+        Raises:
+        - TypeError, if the value is of incorrect type
```

Comment:
Python documentation recommends raising [`TypeError`](https://docs.python.org/2/library/exceptions.html#exceptions.TypeError) when the value is of incorrect type and [`ValueError`](https://docs.python.org/2/library/exceptions.html#exceptions.ValueError) when the value is of correct type but invalid in some other way.

Reply:
However, I am nixing `TypeError`s from coercing, since already `ProcessResult.coerce` violates this recommendation (and I suppose it will not use `TypeError`s anytime soon).